presto-main/src/main/java/com/facebook/presto/memory/ClusterMemoryLeakDetector.java

```diff
@@ -0,0 +1,103 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.memory;
+
+import com.facebook.presto.execution.QueryInfo;
+import com.facebook.presto.spi.QueryId;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Maps;
+import io.airlift.log.Logger;
+import org.joda.time.DateTime;
+
+import javax.annotation.concurrent.GuardedBy;
+import javax.annotation.concurrent.ThreadSafe;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Set;
+import java.util.function.Supplier;
+
+import static com.facebook.presto.execution.QueryState.RUNNING;
+import static com.google.common.collect.ImmutableMap.toImmutableMap;
+import static java.util.Objects.requireNonNull;
+import static org.joda.time.DateTime.now;
+import static org.joda.time.Seconds.secondsBetween;
+
+@ThreadSafe
+public class ClusterMemoryLeakDetector
```

Comment:
cmt msg
> Make ClusterMemoryManager resilient to memory accounting leaks

i would call it `Detect and report memory leaks in ClusterMemoryManager`

To me "making resilient" would hint at "being able to recover automatically".

Reply:
Currently `ClusterMemoryManager` stops functioning when a query is killed and there is a leak for that query. The leak detector should detect that so that `ClusterMemoryManager` should recover from that state.